enhancements/olm/platform-operators.md

```diff
@@ -0,0 +1,545 @@
+
+### The PO Manager (POM)
+
+A PO manager (POM) component will also be introduced, which is responsible for watching and reconciling the PlatformOperator custom resources, and ensuring the set of admin desired OLM-based operators get installed on the cluster. This component will also be responsible for proxying individual PO state back to the CVO.
+
+The CVO has the current limitation where any ClusterOperator resource that is defined outside of the payload will be ignored when determining whether cluster lifecycle events were successful. This means the POM component will need to manage a single, "aggregate" ClusterOperator resource in order to relay state back to the CVO, and result in zero CVO changes needed for this phase.
+
+This "aggregate" ClusterOperator resource will ensure that cluster installations are not considered complete until all the POs are installed on the cluster. Future phases will further iron out this interaction between the CVO components and any OLM-based operators that are managed by this PO mechanism, and extend this interaction to influence cluster upgrades.
+
+When reconciling PlatformOperator custom resources, the POM will reflect any failure states back to the individual PlatformOperator custom resource it's currently reconciling:
+
+- Was the PO manager able to successfully source/find the desired package name?
+- Was the PO manager able to successfully install the desired registry+v1 bundle contents?
+
+At a high-level, this component is responsible for the following:
+
+- Sourcing OLM bundles from supported catalog sources.
+- Applying the sourced bundles, and persisting their manifests to the cluster.
+- Relaying state back to the CVO component through the "core" and "aggregate" ClusterOperator resource.
+
+The implementation of those responsibilities are discussed in more detail below.
+
+### Configuring the "PlatformOperator" Resource
+
+When configuring an individual PlatformOperator resource, the resource's metadata.Name will be used to express the desired PO package name:
+
+```yaml=
+apiVersion: platform.openshift.io/v1alpha1
+kind: PlatformOperator
+metadata:
+  name: openshift-cert-manager-operator
+spec: {}
+```
+
+This approach was favored over an explicit spec field as using the resource's metadata.Name inherits package name uniqueness during phase 0. The POM component would be responsible for finding that package name from OLM's catalog content.
```

Comment:
> When you say during phase 0, does that mean you are intending to change that behaviour later?

Yup, the intention is that this API is limited in surface this phase and it's only major responsibility here is to proxy information back to the CVO. In future phases, where we'll start to look towards building a first-class UX through the installer's install-config API, we'll want to expose more knobs to admins.

> What would happen if 2 POs pointed to the same package name? I think we need to explain in this sentence why that would be bad

The likely issue here is around potential races and non-deterministic behavior. In the case where an individual PO has already sourced and applied one of the bundles in that package. and another PO is configured to use the same desired package name, then we may run into non-determinism when finding candidate bundles in a package:

- A catalog source in the cluster that contains this package got patched, and now new bundle versions are available
- A catalog source in the cluster is down, and we source the desired package from another source

In the case we don't hit non-determinism and the same bundle version is returned in both cases, then one of the PO installations will fail given we already have an underlying BundleDeployment resource managing those bundle contents on the cluster, and this validation behavior is already codified in OLM's rukpak component today.

But if we're using the resource's metadata.Name for configuring the package name, then both of those situations I outlined should be mitigated given we'll get uniqueness of the resource names for free from Kubernetes.

Reply:
> In future phases, where we'll start to look towards building a first-class UX through the installer's install-config API, we'll want to expose more knobs to admins.

The installer config is already a rather large and duplicates a lot of the openshift API already, are the installer team aware of plans to extend it further?

> But if we're using the resource's metadata.Name for configuring the package name, then both of those situations I outlined should be mitigated given we'll get uniqueness of the resource names for free from Kubernetes.

At the risk of duplicating threads, we should probably stop discussing on the other one and continue here. But, you mentioned that the set of characters available for metadata.name isn't sufficient, so we probably can't use metadata.name in the way outlined here. Instead, could webhook validation check for uniqueness? Or some controller later looking at the names and then accepting or rejecting based on uniqueness before anything else runs, we could use a condition for this acceptance

Reply:
> The installer config is already a rather large and duplicates a lot of the openshift API already, are the installer team aware of plans to extend it further?

We haven't had any in-depth conversations yet, and may need to start engaging them as we starting designing phase 1 requirements. @bparees might have a better idea on how we should tackle that.

> Instead, could webhook validation check for uniqueness? Or some controller later looking at the names and then accepting or rejecting based on uniqueness before anything else runs, we could use a condition for this acceptance

This seems like it would also work, and we've discussed this implementation in the past. I'm trying to dig up whether we have any validation logic, or schema definitions for package names in OLM's file-based catalog format to double-check whether this is a more than a potential edge case.

---

And just massaging that other thread here:

> Adding this limitation up front concerns me a little, as, typically when this kind of thing happens, we are reluctant to change or remodel the API later (speaking from reviewing other API evolutions), so, I'm hesitant to say we shouldn't try and think about this ahead of time rather than leaving it til later

I had asked a [similar question before](https://github.com/openshift/enhancements/pull/1170#discussion_r915236030) and it sounds like we can mitigate some of these UX limitations through a combination of documentation, using existing clusters for discoverability, and hooking into gitops workflows to standardize deployments of clusters over time.

Do we have any rope here given we'll be tech preview during phase 0 (and potentially next phase as well) and make any requisite behavior changes in subsequent phases given migration is out-of-scope for us? Typically, it would be ideal to get the API behavior right upfront, but we're essentially rewriting OLM entirely in parallel to this platform operator work, and we may see the top-level PO API be a wrapper around OLM's "Operator" API in the future.

Reply:
There's also a dedicated thread on this top-level PO API that's still unresolved: https://github.com/openshift/enhancements/pull/1170#discussion_r915107267.

Reply:
> We haven't had any in-depth conversations yet, and may need to start engaging them as we starting designing phase 1 requirements. @bparees might have a better idea on how we should tackle that.

this EP process is the way to engage them (tag them and asked them to review the portions of the proposal that impact the install-config api/installer), when we have something that is worth their time to review.

Reply:
Thanks - it sounds like that's a bit premature given we've barely looked at prototyping phase 1 yet. We can look to re-engaging once we have something more concrete in place next phase.

Reply:
Reading through the other thread, I think I should probably raise my concerns about the naming there, so that those who originally suggested reusing the name are tagged

Reply:
Sounds good. Consolidating this thread into the [original](https://github.com/openshift/enhancements/pull/1170#discussion_r935815024). Marking this as resolved in the meantime.